### PR TITLE
Add compatibility with ghc-7.10

### PR DESCRIPTION
**Pull request by qnikst**

### CODE REVIEW

**Comment by qnikst** [comment]
This PR depends on:

https://github.com/haskell-distributed/rank1dynamic/pull/9
https://github.com/haskell-distributed/network-transport-tcp/pull/9
https://github.com/haskell-distributed/network-transport/pull/8

**Comment by qnikst** [comment]
I have splited commits into small one and verified build on 7.8 and 7.10

**Comment by facundominguez** [comment]
Building with ghc-7.4 is failing for some reason.

File: src/Control/Distributed/Process/Internal/Types.hs [hunk 1]
```diff
@@ -137,7 +137,7 @@ import GHC.Generics
 newtype NodeId = NodeId { nodeAddress :: NT.EndPointAddress }
   deriving (Eq, Ord, Typeable, Data, Generic)
 instance Binary NodeId where
-instance NFData NodeId
+instance NFData NodeId where rnf (NodeId a) = rnf a `seq` ()
```
Comment (mboes) [inline src/Control/Distributed/Process/Internal/Types.hs]: We should depend on deepseq>=1.4 and drop the `where rnf (NodeId a) = rnf a seq ()`.

http://hackage.haskell.org/package/deepseq-1.4.1.1/docs/Control-DeepSeq.html#t:NFData

Unless I'm misunderstanding the point of this change?

File: src/Control/Distributed/Process/Internal/Types.hs [hunk 2]
```diff
@@ -497,6 +511,10 @@ data DiedReason =
   | DiedUnknownId
   deriving (Show, Eq)
 
+instance NFData DiedReason where
+  rnf (DiedException s) = rnf s `seq` ()
```
Comment (mboes) [inline src/Control/Distributed/Process/Internal/Types.hs]: can't auto derive?

File: src/Control/Distributed/Process/Internal/Types.hs [hunk 3]
```diff
@@ -366,6 +373,10 @@ data Message =
   }
   deriving (Typeable)
 
+instance NFData Message where
+  rnf (EncodedMessage _ e) = rnf e `seq` ()
+  rnf (UnencodedMessage _ a) = a `seq` ()   -- forced to WHNF only
```
Comment (mboes) [inline src/Control/Distributed/Process/Internal/Types.hs]: why?

File: src/Control/Distributed/Process/Internal/Types.hs (hunk 2 above)
Comment (qnikst) [inline src/Control/Distributed/Process/Internal/Types.hs]: No, autoderiving creates incorrect instances in <GHC-7.10. Those instances will be evaluated to WHNF, not NF.

File: src/Control/Distributed/Process/Internal/Types.hs (hunk 1 above)
Comment (qnikst) [inline src/Control/Distributed/Process/Internal/Types.hs]: There are 2 possibilities, use generic deriving on `>deepseq-1.4` and write instances manually on previous versions, or write instances manually everywhere. I find latter approach more straightforward as former implies latter under if-condition.

File: src/Control/Distributed/Process/Internal/Types.hs (hunk 3 above)
Comment (qnikst) [inline src/Control/Distributed/Process/Internal/Types.hs]: In unencoded message we have existential variable without NFData constraint, so we just can't force it to NF. This semantics was not changed in this pull request. AFAIU the reason for this change is ability to pass messages locally without additional serialization or forcing.

**Comment by qnikst** [comment]
I fixed NFData instance for LazyByteString argument.

**Comment by hyperthunk** [comment]
Can't build this right now and travis is failing. Can anyone confirm it had been fixed for 7.4?

**Comment by qnikst** [comment]
I see "all is well" in travis report:

https://travis-ci.org/haskell-distributed/distributed-process/builds/57053812

@hyperthunk you can run rebuild in top right corner (near settings).

**Comment by hyperthunk** [comment]
Yes will do, I've been away for a month so I'm just catching up again.

**Comment by isaiah** [comment]
Compiling failed on 7.10.1

```
src/Control/Distributed/Process/Internal/Types.hs:141:47:
    No instance for (NFData NT.EndPointAddress)
      arising from a use of ‘rnf’
    In the first argument of ‘seq’, namely ‘rnf a’
    In the expression: rnf a `seq` ()
    In an equation for ‘rnf’: rnf (NodeId a) = rnf a `seq` ()
cabal: Error: some packages failed to install:
distributed-process-0.5.3 failed during the building phase. The exception was:
ExitFailure 1
```

**Comment by qnikst** [comment]
@isaiah have you used patched version of `network-transport`? Almost all parts of distributed-process stack have changes. But I hope we will get over this situation soon.

**Comment by isaiah** [comment]
@qnikst Yes, install `network-transport` from master does fix the problem. Thanks.

**Comment by qnikst** [comment]
This one works good for several people on the zurihack, and passes tests, so I'm going to merge this one.

**Pull request merged**